File: app-server/module-repository/src/test/java/com/epam/esm/gift/repository/TagRepositoryTest.java
package com.epam.esm.gift.repository;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.from;

import java.math.BigDecimal;
import java.util.List;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
import org.springframework.transaction.annotation.Transactional;
import org.testcontainers.junit.jupiter.Testcontainers;

import com.epam.esm.gift.model.Certificate;
import com.epam.esm.gift.model.Tag;

@Transactional
@SpringJUnitConfig(TestDatabaseConfig.class)
@Testcontainers(disabledWithoutDocker = true)
class TagRepositoryTest {

    @Autowired
    private TagRepository repository;

    @Autowired
    private CertificateRepository certificateRepository;

    @Test
    void should_persist_and_find_new_tag() {
        // given
        var initialTag = new Tag();
        initialTag.setName("Test tag");

        // when
        var createdTag = repository.create(initialTag);
        var tagFromRepository = repository.findById(createdTag.getId());

        // then
        assertThat(tagFromRepository)
            .isNotNull()
            .isNotEmpty()
            .contains(initialTag);
    }

    @Test
    void should_update_an_existing_tag() {
        // given
        var initialTag = new Tag();
        initialTag.setName("Test tag");
        var updatedName = "Updated test tag name";

        // when
        var createdTag = repository.create(initialTag);
        var updatedTag = repository.update(createdTag.copy().name(updatedName).build());

        // then
        assertThat(updatedTag.getName())
            .isNotNull()
            .isNotEqualTo(createdTag.getName())
            .isEqualTo(updatedName);
    }

    @Test
    void should_return_list_of_all_existing_tags() {
        // given
        var initialTags = List.of(
            Tag.builder().name("Test tag #1").build(),
            Tag.builder().name("Test tag #2").build(),
            Tag.builder().name("Test tag #3").build()
        );

        initialTags.forEach(repository::create);

        // when
        var allExistingTags = repository.findAll();

        // then
        assertThat(allExistingTags)
            .isNotNull()
            .hasSameSizeAs(initialTags)
            .containsAll(initialTags);
    }

    @Test
    void should_delete_existing_tag() {
        // given
        var initialTag = Tag.builder().name("Test tag").build();

        // when
        var createdTag = repository.create(initialTag);
        repository.delete(createdTag);
        var tagFromRepository = repository.findById(createdTag.getId());

        // then
        assertThat(tagFromRepository)
            .isNotNull()
            .isEmpty();
    }

    @Test
    void should_delete_existing_tag_by_its_id() {
        // given
        var initialTag = Tag.builder().name("Test tag").build();

        // when
        var createdTag = repository.create(initialTag);
        repository.deleteById(createdTag.getId());
        var tagFromRepository = repository.findById(createdTag.getId());

        // then
        assertThat(tagFromRepository)
            .isNotNull()
            .isEmpty();
    }

    @Test
    void should_add_tags_to_an_existing_certificate() {
        // given
        var initialTags = List.of(
            Tag.builder().name("Test tag #1").build(),
            Tag.builder().name("Test tag #2").build(),
            Tag.builder().name("Test tag #3").build()
        );

        for (var initialTag : initialTags) {
            repository.create(initialTag);
        }

        var certificate = certificateRepository.create(
            Certificate.builder()
                .name("Test certificate name")
                .description("Test certificate description")
                .price(new BigDecimal("1.00"))
                .duration(30)
                .build()
        );

        // when
        repository.addTagsToCertificate(certificate.getId(), initialTags);
        var tagsByCertificateId = repository.findTagsByCertificateId(certificate.getId());

        // then
        assertThat(tagsByCertificateId)
            .isNotNull()
            .isEqualTo(initialTags);
    }

    @Test
    void should_remove_tags_from_an_existing_certificate() {
        // given
        var initialTags = List.of(
            Tag.builder().name("Test tag #1").build(),
            Tag.builder().name("Test tag #2").build(),
            Tag.builder().name("Test tag #3").build()
        );

        initialTags.forEach(repository::create);

        var certificate = certificateRepository.create(
            Certificate.builder()
                .name("Test certificate name")
                .description("Test certificate description")
                .price(new BigDecimal("1.00"))
                .duration(30)
                .build()
        );

        // when
        repository.addTagsToCertificate(certificate.getId(), initialTags);
        repository.removeAllTagsFromCertificate(certificate.getId());
        var tagsByCertificateId = repository.findTagsByCertificateId(certificate.getId());

        // then
        assertThat(tagsByCertificateId)
            .isNotNull()
            .isEmpty();
    }

    @Test
    void should_find_tag_by_its_name() {
        // given
        var expectedName = "Test tag #1";

        repository.create(Tag.builder().name(expectedName).build());

        // when
        var tag = repository.findTagByName(expectedName);

        // then
        assertThat(tag)
            .isNotNull()
            .isPresent()
            .get()
            .returns(expectedName, from(Tag::getName));
    }

    @Test
    void should_find_tags_by_their_names() {
        // given
        var initialTagNames = List.of(
            "Test tag #1",
            "Test tag #2",
            "Test tag #3"
        );

        initialTagNames.stream()
            .map(name -> Tag.builder().name(name).build())
            .forEach(repository::create);

        // when
        var tags = repository.findTagByNames(initialTagNames);

        // then
        assertThat(tags)
            .isNotNull()
            .hasSameSizeAs(initialTagNames)
            .allMatch(tag -> initialTagNames.contains(tag.getName()));
    }
}

File: app-server/module-repository/src/test/kotlin/com/epam/esm/gift/repository/TagRepositoryTest.kt
package com.epam.esm.gift.repository

import com.epam.esm.gift.model.Certificate
import com.epam.esm.gift.model.Tag
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.from
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig
import org.springframework.transaction.annotation.Transactional
import org.testcontainers.junit.jupiter.Testcontainers
import java.math.BigDecimal
import org.junit.jupiter.api.Tag as TestTag

@TestTag("unit")
@Transactional
@SpringJUnitConfig(TestDatabaseConfig::class)
@Testcontainers(disabledWithoutDocker = true)
internal class TagRepositoryTest {

    @Autowired
    private lateinit var repository: TagRepository

    @Autowired
    private lateinit var certificateRepository: CertificateRepository

    @Test
    fun `should persist and find new tag`() {
        // given
        val initialTag = Tag().apply { name = "Test tag" }

        // when
        val createdTag = repository.create(initialTag)
        val tagFromRepository = repository.findById(createdTag.id)

        // then
        assertThat(tagFromRepository)
            .isNotNull
            .isNotEmpty
            .contains(initialTag)
    }

    @Test
    fun `should update an existing tag`() {
        // given
        val initialTag = Tag().apply { name = "Test tag" }
        val updatedName = "Updated test tag name"

        // when
        val createdTag = repository.create(initialTag)
        val updatedTag = repository.update(createdTag.copy().name(updatedName).build())

        // then
        assertThat(updatedTag.name)
            .isNotNull
            .isNotEqualTo(createdTag.name)
            .isEqualTo(updatedName)
    }

    @Test
    fun `should return list of all existing tags`() {
        // given
        val initialTags = listOf(
            Tag.builder().name("Test tag #1").build(),
            Tag.builder().name("Test tag #2").build(),
            Tag.builder().name("Test tag #3").build(),
        )

        initialTags.forEach(repository::create)

        // when
        val allExistingTags = repository.findAll()

        // then
        assertThat(allExistingTags)
            .isNotNull
            .hasSameSizeAs(initialTags)
            .containsAll(initialTags)
    }

    @Test
    fun `should delete existing tag`() {
        // given
        val initialTag = Tag.builder().name("Test tag").build()

        // when
        val createdTag = repository.create(initialTag)
        repository.delete(createdTag)
        val tagFromRepository = repository.findById(createdTag.id)

        // then
        assertThat(tagFromRepository)
            .isNotNull
            .isEmpty
    }

    @Test
    fun `should delete existing tag by its id`() {
        // given
        val initialTag = Tag.builder().name("Test tag").build()

        // when
        val createdTag = repository.create(initialTag)
        repository.deleteById(createdTag.id)
        val tagFromRepository = repository.findById(createdTag.id)

        // then
        assertThat(tagFromRepository)
            .isNotNull
            .isEmpty
    }

    @Test
    fun `should add tags to an existing certificate`() {
        // given
        val initialTags = listOf(
            Tag.builder().name("Test tag #1").build(),
            Tag.builder().name("Test tag #2").build(),
            Tag.builder().name("Test tag #3").build()
        )

        for (initialTag in initialTags) {
            repository.create(initialTag)
        }

        val certificate = certificateRepository.create(
            Certificate.builder()
                .name("Test certificate name")
                .description("Test certificate description")
                .price(BigDecimal("1.00"))
                .duration(30)
                .build()
        )

        // when
        repository.addTagsToCertificate(certificate.id, initialTags)
        val tagsByCertificateId = repository.findTagsByCertificateId(certificate.id)

        // then
        assertThat(tagsByCertificateId)
            .isNotNull
            .isEqualTo(initialTags)
    }

    @Test
    fun `should remove tags from an existing certificate`() {
        // given
        val initialTags = listOf(
            Tag.builder().name("Test tag #1").build(),
            Tag.builder().name("Test tag #2").build(),
            Tag.builder().name("Test tag #3").build()
        )

        initialTags.forEach(repository::create)

        val certificate = certificateRepository.create(
            Certificate.builder()
                .name("Test certificate name")
                .description("Test certificate description")
                .price(BigDecimal("1.00"))
                .duration(30)
                .build()
        )

        // when
        repository.addTagsToCertificate(certificate.id, initialTags)
        repository.removeAllTagsFromCertificate(certificate.id)
        val tagsByCertificateId = repository.findTagsByCertificateId(certificate.id)

        // then
        assertThat(tagsByCertificateId)
            .isNotNull
            .isEmpty()
    }

    @Test
    fun `should find tag by its name`() {
        // given
        val expectedName = "Test tag #1"

        repository.create(Tag().apply { name = expectedName })

        // when
        val tag = repository.findTagByName(expectedName)

        // then
        assertThat(tag)
            .isNotNull
            .isPresent
            .get()
            .returns(expectedName, from(Tag::getName))
    }

    @Test
    fun `should find tags by their names`() {
        // given
        val initialTagNames = listOf(
            "Test tag #1",
            "Test tag #2",
            "Test tag #3"
        )

        initialTagNames.map { Tag().apply { name = it } }.forEach(repository::create)

        // when
        val tags = repository.findTagByNames(initialTagNames)

        // then
        assertThat(tags)
            .isNotNull
            .hasSameSizeAs(initialTagNames)
            .allMatch { tag -> initialTagNames.contains(tag.name) }
    }
}